[ROCm] Porting LoadOp and StoreOp for AMDGPU

B1tway opened the pull request:
Tasks:
1) Implement GCNBuilder *[Finished]*
2) Implementation of helpers for AMDGCN memory instructions *[Finished]*
3) Porting triton::StoreOp *[Finished]*
4)  Porting triton::LoadOp *[Finished]*

I still decided on the inline asm solution, because at the moment it is much easier to integrate into Triton than the `memref.load` solution. 
I think in the future it will be possible to redo the LoadOp and StoreOp part with the memref dialect and maybe that will be a good solution for Nvidia too

codego7250 commented on include/triton/Conversion/TritonGPUToLLVM/AsmFormat.h:
```diff
@@ -0,0 +1,33 @@
+  std::string osStr;
+  llvm::raw_string_ostream os(osStr);
+  for (size_t i = 0; !strs.empty() && i < strs.size() - 1; ++i)
+    os << strs[i] << delimiter;
```
The final string is not joined. Is this the right behavior? 
If yes, you may want to add a comment.

rsanthanam-amd replied:
we moved this existing function from PtxAsmFormat.cpp to a common header.

unfortunately, it is not clear to me whether this is a bug or expected behavior.

once we have more functionality added to the PR, then I think we can test it to verify.

thank you for pointing it out!

codego7250 commented:
> 

GCN inlining is the right infrastructure to move forward. Not only for rapid iteration, It maybe a must since you may want to implement the shape manipulation and etc. in assembly which is not exposed as a ISA primitives.

B1tway commented:
I added support for modifiers in AMDGCN inline asm.  For example, now we can do memory instructions with `offset` modifier.

I also plan to add tests similar to PtxAsmFormatTest for GCNBuilder in the near future

rsanthanam-amd commented:
@B1tway that's great!

do we have enough here to implement TritonGPU LoadOP?

B1tway commented:
@rsanthanam-amd 
I think so. I plan to finish this PR this weekend.

rsanthanam-amd commented:
awesome.  can't wait to see it.

thanks!

B1tway commented:
Input MLIR kernel
```

B1tway commented:
Input MLIR kernel:
```
#blocked0 = #triton_gpu.blocked<{sizePerThread = [2], threadsPerWarp = [64], warpsPerCTA = [2], order = [0]}>
module attributes {"triton_gpu.num-warps" = 2 : i32} {
  func @global_load_store(%arg0: !tt.ptr<f32> {tt.divisibility = 4 : i32}, %arg1: !tt.ptr<f32> {tt.divisibility = 4 : i32}, %arg2: !tt.ptr<f32> {tt.divisibility = 4 : i32}, %arg3: i32) {
    %c256_i32 = arith.constant 256 : i32
    %0 = tt.get_program_id {axis = 0 : i32} : i32
    %1 = arith.muli %0, %c256_i32 : i32
    %2 = tt.make_range {end = 256 : i32, start = 0 : i32} : tensor<256xi32, #blocked0>
    %3 = tt.splat %1 : (i32) -> tensor<256xi32, #blocked0>
    %4 = arith.addi %3, %2 : tensor<256xi32, #blocked0>
    %5 = tt.splat %arg0 : (!tt.ptr<f32>) -> tensor<256x!tt.ptr<f32>, #blocked0>
    %6 = tt.addptr %5, %4 : tensor<256x!tt.ptr<f32>, #blocked0>
    %7 = tt.splat %arg1 : (!tt.ptr<f32>) -> tensor<256x!tt.ptr<f32>, #blocked0>
    %8 = tt.addptr %7, %4 : tensor<256x!tt.ptr<f32>, #blocked0>


    %9 = tt.load %6 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<256xf32, #blocked0>
    %10 = tt.load %8 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<256xf32, #blocked0>
    %11 = arith.addf %9, %10 : tensor<256xf32, #blocked0>
    %12 = tt.splat %arg2 : (!tt.ptr<f32>) -> tensor<256x!tt.ptr<f32>, #blocked0>
    %13 = tt.addptr %12, %4 : tensor<256x!tt.ptr<f32>, #blocked0>

    tt.store %13, %11 : tensor<256xf32, #blocked0>
    return
  }
```
Output AMDGCN kernel:
```
        .text
        .amdgcn_target "amdgcn-amd-amdhsa--gfx906:sramecc+:xnack-"
        .globl  global_load_store
        .p2align        8
        .type   global_load_store,@function
global_load_store:
.Lfunc_begin0:
        .file   1 "/git/triton-rocm-fork/test/Target/tritongpu_to_ptx.mlir"
        .loc    1 9 0
        .cfi_sections .debug_frame
        .cfi_startproc
        v_writelane_b32 v8, s6, 0
        s_mov_b64 s[6:7], s[4:5]
        v_readlane_b32 s4, v8, 0
.Ltmp0:
        .loc    1 11 10 prologue_end
        s_load_dwordx2 s[10:11], s[6:7], 0x0
        s_load_dwordx2 s[8:9], s[6:7], 0x8
        s_load_dwordx2 s[6:7], s[6:7], 0x10
        s_mov_b32 s5, 1
        .loc    1 13 10
        v_lshlrev_b32_e64 v0, s5, v0
        s_mov_b32 s5, 0xfe
        v_and_b32_e64 v1, v0, s5
        s_mov_b32 s5, 8
        .loc    1 15 10
        v_mov_b32_e32 v0, s5
        v_lshl_or_b32 v0, s4, v0, v1
        .loc    1 17 10
        v_ashrrev_i32_e64 v2, 31, v0
        v_mov_b32_e32 v1, v2
        s_mov_b32 s4, 2
        v_lshlrev_b64 v[6:7], s4, v[0:1]
        s_waitcnt lgkmcnt(0)
        s_mov_b32 s4, s10
        v_mov_b32_e32 v0, v6
        s_mov_b32 s10, s11
        v_mov_b32_e32 v2, v7
        v_add_co_u32_e64 v0, s[4:5], s4, v0
        v_mov_b32_e32 v1, s10
        v_addc_co_u32_e64 v2, s[4:5], v1, v2, s[4:5]
        v_mov_b32_e32 v1, v2
        .loc    1 22 10
        ;;#ASMSTART
        global_load_dword v0, v[0:1] offset:0;
        global_load_dword v1, v[0:1] offset:4;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        .loc    1 19 10
        s_mov_b32 s4, s8
        v_mov_b32_e32 v2, v6
        s_mov_b32 s8, s9
        v_mov_b32_e32 v4, v7
        v_add_co_u32_e64 v2, s[4:5], s4, v2
        v_mov_b32_e32 v3, s8
        v_addc_co_u32_e64 v4, s[4:5], v3, v4, s[4:5]
        v_mov_b32_e32 v3, v4
        .loc    1 23 11
        ;;#ASMSTART
        global_load_dword v5, v[2:3] offset:0;
        global_load_dword v4, v[2:3] offset:4;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        .loc    1 26 11
        s_mov_b32 s4, s6
        v_mov_b32_e32 v2, v6
        s_mov_b32 s6, s7
        v_mov_b32_e32 v6, v7
        v_add_co_u32_e64 v2, s[4:5], s4, v2
        v_mov_b32_e32 v3, s6
        v_addc_co_u32_e64 v6, s[4:5], v3, v6, s[4:5]
        v_mov_b32_e32 v3, v6
        .loc    1 24 11
        v_add_f32_e64 v0, v0, v5
        v_add_f32_e64 v1, v1, v4
        .loc    1 28 5
        ;;#ASMSTART
        global_store_dword v[2:3], v0 offset:0;
        global_store_dword v[2:3], v1 offset:4;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        .loc    1 29 5
        s_endpgm
.Ltmp1:
        .section        .rodata,#alloc
        .p2align        6
        .amdhsa_kernel global_load_store
                .amdhsa_group_segment_fixed_size 0
                .amdhsa_private_segment_fixed_size 0
                .amdhsa_kernarg_size 28
                .amdhsa_user_sgpr_count 6
                .amdhsa_user_sgpr_private_segment_buffer 1
                .amdhsa_user_sgpr_dispatch_ptr 0
                .amdhsa_user_sgpr_queue_ptr 0
                .amdhsa_user_sgpr_kernarg_segment_ptr 1
                .amdhsa_user_sgpr_dispatch_id 0
                .amdhsa_user_sgpr_flat_scratch_init 0
                .amdhsa_user_sgpr_private_segment_size 0
                .amdhsa_system_sgpr_private_segment_wavefront_offset 0
                .amdhsa_system_sgpr_workgroup_id_x 1
                .amdhsa_system_sgpr_workgroup_id_y 0
                .amdhsa_system_sgpr_workgroup_id_z 0
                .amdhsa_system_sgpr_workgroup_info 0
                .amdhsa_system_vgpr_workitem_id 0
                .amdhsa_next_free_vgpr 9
                .amdhsa_next_free_sgpr 12
                .amdhsa_reserve_vcc 0
                .amdhsa_reserve_flat_scratch 0
                .amdhsa_reserve_xnack_mask 0
                .amdhsa_float_round_mode_32 0
                .amdhsa_float_round_mode_16_64 0
                .amdhsa_float_denorm_mode_32 3
                .amdhsa_float_denorm_mode_16_64 3
                .amdhsa_dx10_clamp 1
                .amdhsa_ieee_mode 1
                .amdhsa_fp16_overflow 0
                .amdhsa_exception_fp_ieee_invalid_op 0
                .amdhsa_exception_fp_denorm_src 0
                .amdhsa_exception_fp_ieee_div_zero 0
                .amdhsa_exception_fp_ieee_overflow 0
                .amdhsa_exception_fp_ieee_underflow 0
                .amdhsa_exception_fp_ieee_inexact 0
                .amdhsa_exception_int_div_zero 0
        .end_amdhsa_kernel
        .text
.Lfunc_end0:
        .size   global_load_store, .Lfunc_end0-global_load_store
        .cfi_endproc

        .section        .debug_abbrev
        .byte   1
        .byte   17
        .byte   1
        .byte   37
        .byte   14
        .byte   19
        .byte   5
        .byte   3
        .byte   14
        .byte   16
        .byte   23
        .byte   27
        .byte   14
        .ascii  "\264B"
        .byte   25
        .byte   17
        .byte   1
        .byte   18
        .byte   6
        .byte   0
        .byte   0
        .byte   2
        .byte   46
        .byte   0
        .byte   17
        .byte   1
        .byte   18
        .byte   6
        .byte   110
        .byte   14
        .byte   3
        .byte   14
        .byte   58
        .byte   11
        .byte   59
        .byte   11
        .byte   63
        .byte   25
        .byte   0
        .byte   0
        .byte   0
        .section        .debug_info
.Lcu_begin0:
        .long   .Ldebug_info_end0-.Ldebug_info_start0
.Ldebug_info_start0:
        .short  4
        .long   .debug_abbrev
        .byte   8
        .byte   1
        .long   .Linfo_string0
        .short  2
        .long   .Linfo_string1
        .long   .Lline_table_start0
        .long   .Linfo_string2

        .quad   .Lfunc_begin0
        .long   .Lfunc_end0-.Lfunc_begin0
        .byte   2
        .quad   .Lfunc_begin0
        .long   .Lfunc_end0-.Lfunc_begin0
        .long   .Linfo_string3
        .long   .Linfo_string3
        .byte   1
        .byte   9

        .byte   0
.Ldebug_info_end0:
        .section        .debug_str,"MS",@progbits,1
.Linfo_string0:
        .asciz  "mlir"
.Linfo_string1:
        .asciz  "LLVMDialectModule"
.Linfo_string2:
        .asciz  "/"
.Linfo_string3:
        .asciz  "global_load_store"
        .section        .debug_pubnames
        .long   .LpubNames_end0-.LpubNames_start0
.LpubNames_start0:
        .short  2
        .long   .Lcu_begin0
        .long   66
        .long   42
        .asciz  "global_load_store"
        .long   0
.LpubNames_end0:
        .section        .debug_pubtypes
        .long   .LpubTypes_end0-.LpubTypes_start0
.LpubTypes_start0:
        .short  2
        .long   .Lcu_begin0
        .long   66
        .long   0
.LpubTypes_end0:
        .section        ".note.GNU-stack"
        .amdgpu_metadata
---
amdhsa.kernels:
  - .args:
      - .address_space:  global
        .offset:         0
        .size:           8
        .value_kind:     global_buffer
      - .address_space:  global
        .offset:         8
        .size:           8
        .value_kind:     global_buffer
      - .address_space:  global
        .offset:         16
        .size:           8
        .value_kind:     global_buffer
      - .offset:         24
        .size:           4
        .value_kind:     by_value
    .group_segment_fixed_size: 0
    .kernarg_segment_align: 8
    .kernarg_segment_size: 28
    .max_flat_workgroup_size: 1024
    .name:           global_load_store
    .private_segment_fixed_size: 0
    .sgpr_count:     12
    .sgpr_spill_count: 1
    .symbol:         global_load_store.kd
    .vgpr_count:     9
    .vgpr_spill_count: 0
    .wavefront_size: 64
amdhsa.target:   'amdgcn-amd-amdhsa--gfx906:sramecc+:xnack-'
amdhsa.version:
  - 1
  - 1
...

        .end_amdgpu_metadata
        .section        .debug_line
.Lline_table_start0:

```

rsanthanam-amd commented:
@B1tway 

Thanks.

I am unable to reproduce this result because I am getting errors.

I noticed that you have not updated your branch to the required upstream commit (e517b58d59ba96357d042d8fa5819a690d00d749).

It seems like there are differences between the version of `lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp` that is currently in your branch and the version that exists in the upstream commit cited above that could be causing issues with your current changes.

rsanthanam-amd commented:
@B1tway 

Thanks.  I was able to successfully reproduce this result with the float32 data type.

B1tway commented:
@rsanthanam-amd 
This PR works and can be reviewed. I will add support for int8 in a separate PR

codego7250 reviewed (approved):
Pls enable the testcases in testcore for load/store to show the "global.load" instead of "load.global" for ptx.

codego7250 commented:
> @rsanthanam-amd This PR works and can be reviewed. I will add support for int8 in a separate PR

Just confirm that fp16/bf16 will work, right?

codego7250 commented on lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp:
```diff
@@ -883,6 +884,39 @@ struct LoadOpConversion
+
+      for (size_t wordIdx = 0; wordIdx < nWords; ++wordIdx) {
+        auto &gload =
+            gcnBuilder.create<GCNMemInstr>("global_load")->type(width);
```
In GCN, there is V4 version of load/store. Will this width reflect that?

B1tway replied:
In the current version memory load instructions are not vectorized. I plan to do the optimizations a little later

B1tway commented:
Input MLIR kernel:
```
#blocked0 = #triton_gpu.blocked<{sizePerThread = [2], threadsPerWarp = [64], warpsPerCTA = [2], order = [0]}>
module attributes {"triton_gpu.num-warps" = 2 : i32} {
  func @global_load_store(%arg0: !tt.ptr<i8> {tt.divisibility = 4 : i32}, %arg1: !tt.ptr<i8> {tt.divisibility = 4 : i32}, %arg2: !tt.ptr<i8> {tt.divisibility = 4 : i32}, %arg3: i32) {
    %c256_i32 = arith.constant 256 : i32
    %0 = tt.get_program_id {axis = 0 : i32} : i32
    %1 = arith.muli %0, %c256_i32 : i32
    %2 = tt.make_range {end = 256 : i32, start = 0 : i32} : tensor<256xi32, #blocked0>
    %3 = tt.splat %1 : (i32) -> tensor<256xi32, #blocked0>
    %4 = arith.addi %3, %2 : tensor<256xi32, #blocked0>
    %5 = tt.splat %arg0 : (!tt.ptr<i8>) -> tensor<256x!tt.ptr<i8>, #blocked0>
    %6 = tt.addptr %5, %4 : tensor<256x!tt.ptr<i8>, #blocked0>
    %7 = tt.splat %arg1 : (!tt.ptr<i8>) -> tensor<256x!tt.ptr<i8>, #blocked0>
    %8 = tt.addptr %7, %4 : tensor<256x!tt.ptr<i8>, #blocked0>


    %9 = tt.load %6 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<256xi8, #blocked0>
    %10 = tt.load %8 {cache = 1 : i32, evict = 1 : i32, isVolatile = false} : tensor<256xi8, #blocked0>
    %11 = arith.addi %9, %10 : tensor<256xi8, #blocked0>
    %12 = tt.splat %arg2 : (!tt.ptr<i8>) -> tensor<256x!tt.ptr<i8>, #blocked0>
    %13 = tt.addptr %12, %4 : tensor<256x!tt.ptr<i8>, #blocked0>

    tt.store %13, %11 : tensor<256xi8, #blocked0>
    return
  }
}
```
Output AMDGCN kernel:
```
        .text
        .amdgcn_target "amdgcn-amd-amdhsa--gfx906:sramecc+:xnack-"
        .globl  global_load_store
        .p2align        8
        .type   global_load_store,@function
global_load_store:
.Lfunc_begin0:
        .file   1 "/git/triton-rocm-fork/test/Target/tritongpu_to_ptx.mlir"
        .loc    1 9 0
        .cfi_sections .debug_frame
        .cfi_startproc
        v_writelane_b32 v8, s6, 0
        s_mov_b64 s[8:9], s[4:5]
        v_readlane_b32 s4, v8, 0
.Ltmp0:
        .loc    1 11 10 prologue_end
        s_load_dwordx2 s[6:7], s[8:9], 0x0
        s_load_dwordx2 s[10:11], s[8:9], 0x8
        s_load_dwordx2 s[8:9], s[8:9], 0x10
        s_mov_b32 s5, 1
        .loc    1 13 10
        v_lshlrev_b32_e64 v0, s5, v0
        s_mov_b32 s5, 0xfe
        v_and_b32_e64 v1, v0, s5
        s_mov_b32 s5, 8
        .loc    1 15 10
        v_mov_b32_e32 v0, s5
        v_lshl_or_b32 v6, s4, v0, v1
        .loc    1 17 10
        v_ashrrev_i32_e64 v0, 31, v6
        v_mov_b32_e32 v7, v0
        s_waitcnt lgkmcnt(0)
        s_mov_b32 s4, s6
        v_mov_b32_e32 v0, v6
        s_mov_b32 s6, s7
        v_mov_b32_e32 v2, v7
        v_add_co_u32_e64 v0, s[4:5], s4, v0
        v_mov_b32_e32 v1, s6
        v_addc_co_u32_e64 v2, s[4:5], v1, v2, s[4:5]
        v_mov_b32_e32 v1, v2
        .loc    1 22 10
        ;;#ASMSTART
        global_load_byte v5, v[0:1] offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        s_mov_b32 s4, 8
        v_lshrrev_b16_e64 v4, s4, v5
        .loc    1 19 10
        s_mov_b32 s6, s10
        v_mov_b32_e32 v0, v6
        s_mov_b32 s5, s11
        v_mov_b32_e32 v2, v7
        v_add_co_u32_e64 v1, s[6:7], s6, v0
        v_mov_b32_e32 v0, s5
        v_addc_co_u32_e64 v0, s[6:7], v0, v2, s[6:7]
        v_mov_b32_e32 v2, v0
        .loc    1 23 11
        ;;#ASMSTART
        global_load_byte v0, v[1:2] offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        v_lshrrev_b16_e64 v3, s4, v0
        .loc    1 26 11
        s_mov_b32 s6, s8
        v_mov_b32_e32 v1, v6
        s_mov_b32 s5, s9
        v_mov_b32_e32 v6, v7
        v_add_co_u32_e64 v1, s[6:7], s6, v1
        v_mov_b32_e32 v2, s5
        v_addc_co_u32_e64 v6, s[6:7], v2, v6, s[6:7]
        v_mov_b32_e32 v2, v6
        .loc    1 24 11
        v_add_u16_e64 v0, v0, v5
        s_mov_b32 s5, 0xff
        v_and_b32_e64 v0, v0, s5
        v_add_u16_e64 v3, v3, v4
        v_lshlrev_b16_e64 v3, s4, v3
        v_or_b32_e64 v0, v0, v3
        .loc    1 28 5
        s_mov_b32 s4, 0xffff
        v_and_b32_e64 v0, s4, v0
        ;;#ASMSTART
        global_store_byte v[1:2], v0 offset:0;
        s_waitcnt vmcnt(0)  ;
        ;;#ASMEND
        .loc    1 29 5
        s_endpgm
.Ltmp1:
        .section        .rodata,#alloc
        .p2align        6
        .amdhsa_kernel global_load_store
                .amdhsa_group_segment_fixed_size 0
                .amdhsa_private_segment_fixed_size 0
                .amdhsa_kernarg_size 28
                .amdhsa_user_sgpr_count 6
                .amdhsa_user_sgpr_private_segment_buffer 1
                .amdhsa_user_sgpr_dispatch_ptr 0
                .amdhsa_user_sgpr_queue_ptr 0
                .amdhsa_user_sgpr_kernarg_segment_ptr 1
                .amdhsa_user_sgpr_dispatch_id 0
                .amdhsa_user_sgpr_flat_scratch_init 0
                .amdhsa_user_sgpr_private_segment_size 0
                .amdhsa_system_sgpr_private_segment_wavefront_offset 0
                .amdhsa_system_sgpr_workgroup_id_x 1
                .amdhsa_system_sgpr_workgroup_id_y 0
                .amdhsa_system_sgpr_workgroup_id_z 0
                .amdhsa_system_sgpr_workgroup_info 0
                .amdhsa_system_vgpr_workitem_id 0
                .amdhsa_next_free_vgpr 9
                .amdhsa_next_free_sgpr 12
                .amdhsa_reserve_vcc 0
                .amdhsa_reserve_flat_scratch 0
                .amdhsa_reserve_xnack_mask 0
                .amdhsa_float_round_mode_32 0
                .amdhsa_float_round_mode_16_64 0
                .amdhsa_float_denorm_mode_32 3
                .amdhsa_float_denorm_mode_16_64 3
                .amdhsa_dx10_clamp 1
                .amdhsa_ieee_mode 1
                .amdhsa_fp16_overflow 0
                .amdhsa_exception_fp_ieee_invalid_op 0
                .amdhsa_exception_fp_denorm_src 0
                .amdhsa_exception_fp_ieee_div_zero 0
                .amdhsa_exception_fp_ieee_overflow 0
                .amdhsa_exception_fp_ieee_underflow 0
                .amdhsa_exception_fp_ieee_inexact 0
                .amdhsa_exception_int_div_zero 0
        .end_amdhsa_kernel
        .text
.Lfunc_end0:
        .size   global_load_store, .Lfunc_end0-global_load_store
        .cfi_endproc

        .section        .debug_abbrev
        .byte   1
        .byte   17
        .byte   1
        .byte   37
        .byte   14
        .byte   19
        .byte   5
        .byte   3
        .byte   14
        .byte   16
        .byte   23
        .byte   27
        .byte   14
        .ascii  "\264B"
        .byte   25
        .byte   17
        .byte   1
        .byte   18
        .byte   6
        .byte   0
        .byte   0
        .byte   2
        .byte   46
        .byte   0
        .byte   17
        .byte   1
        .byte   18
        .byte   6
        .byte   110
        .byte   14
        .byte   3
        .byte   14
        .byte   58
        .byte   11
        .byte   59
        .byte   11
        .byte   63
        .byte   25
        .byte   0
        .byte   0
        .byte   0
        .section        .debug_info
.Lcu_begin0:
        .long   .Ldebug_info_end0-.Ldebug_info_start0
.Ldebug_info_start0:
        .short  4
        .long   .debug_abbrev
        .byte   8
        .byte   1
        .long   .Linfo_string0
        .short  2
        .long   .Linfo_string1
        .long   .Lline_table_start0
        .long   .Linfo_string2

        .quad   .Lfunc_begin0
        .long   .Lfunc_end0-.Lfunc_begin0
        .byte   2
        .quad   .Lfunc_begin0
        .long   .Lfunc_end0-.Lfunc_begin0
        .long   .Linfo_string3
        .long   .Linfo_string3
        .byte   1
        .byte   9

        .byte   0
.Ldebug_info_end0:
        .section        .debug_str,"MS",@progbits,1
.Linfo_string0:
        .asciz  "mlir"
.Linfo_string1:
        .asciz  "LLVMDialectModule"
.Linfo_string2:
        .asciz  "/"
.Linfo_string3:
        .asciz  "global_load_store"
        .section        .debug_pubnames
        .long   .LpubNames_end0-.LpubNames_start0
.LpubNames_start0:
        .short  2
        .long   .Lcu_begin0
        .long   66
        .long   42
        .asciz  "global_load_store"
        .long   0
.LpubNames_end0:
        .section        .debug_pubtypes
        .long   .LpubTypes_end0-.LpubTypes_start0
.LpubTypes_start0:
        .short  2
        .long   .Lcu_begin0
        .long   66
        .long   0
.LpubTypes_end0:
        .section        ".note.GNU-stack"
        .amdgpu_metadata
---
amdhsa.kernels:
  - .args:
      - .address_space:  global
        .offset:         0
        .size:           8
        .value_kind:     global_buffer
      - .address_space:  global
        .offset:         8
        .size:           8
        .value_kind:     global_buffer
      - .address_space:  global
        .offset:         16
        .size:           8
        .value_kind:     global_buffer
      - .offset:         24
        .size:           4
        .value_kind:     by_value
    .group_segment_fixed_size: 0
    .kernarg_segment_align: 8
    .kernarg_segment_size: 28
    .max_flat_workgroup_size: 1024
    .name:           global_load_store
    .private_segment_fixed_size: 0
    .sgpr_count:     12
    .sgpr_spill_count: 1
    .symbol:         global_load_store.kd
    .vgpr_count:     9
    .vgpr_spill_count: 0
    .wavefront_size: 64
amdhsa.target:   'amdgcn-amd-amdhsa--gfx906:sramecc+:xnack-'
amdhsa.version:
  - 1
  - 1
...

        .end_amdgpu_metadata
        .section        .debug_line
.Lline_table_start0:
```

rsanthanam-amd commented:
@B1tway 

Yes, I tried the changes in my branch which can generate HSACO binaries.

I used it to run test_bin_op '+' operator with int8 operands and unfortunately it did not work for that kernel.

This is the error i am getting:

error: couldn't allocate output register for constraint 'v'

dfukalov commented:
Hi @rsanthanam-amd, I guess the issue is in the generated asm code "global_load_byte $0, $1, **off offset**:0" - I indicated the **incorrect part as bold**. I've just double-checked your branch "triton-mlir-rohit-test" (that if I understand correctly, contains the PR) and found the lines generating the issue:
```
gcnBuilder.newModifier("off offset", std::to_string(offset));
```
but the PR has it as
```
gcnBuilder.newModifier("offset", std::to_string(offset));
```
So I guess that was some merging issue.

Would you please re-check the PR on a vanilla triton-mlir branch?

rsanthanam-amd commented:
@dfukalov 

The **off offset** was intentional.  Based on my research, that is the correct way to express this in AMDGCN.

I had to make that change to get float32 and other data types to work correctly (in other words to successfully generate HSACO).

Going forward, it will be very important for you and your team to be able to generate HSACO as a means to verify that the AMDGCN is correct.

dfukalov commented:
@rsanthanam-amd I'm sorry I expressed my though very unclear. I guessed that we should check the PR on the original "triton-lmir" branch since the PR is for the branch.

Regarding the HSACO generation - we are working on the task and your branch is helpful.

But it seems the PR' changes from its description can be checked with just dumping generated assembly, e.g. with help of functionality introduced in https://github.com/ROCmSoftwarePlatform/triton/pull/25, what do you think?

B1tway commented:
@rsanthanam-amd 
I checked your branch and really found a few bugs.  Thank you!

I've already submitted a commit with the fixes, as well as make a [branch](https://github.com/B1tway/triton/tree/hsaco-testing) based on this PR and your branch where HSACO is generated.

rsanthanam-amd commented on include/triton/Conversion/TritonGPUToLLVM/AsmFormat.h:
```diff
@@ -0,0 +1,33 @@
+namespace triton {
+using llvm::StringRef;
+
+static std::string strJoin(llvm::ArrayRef<std::string> strs,
```
I would prefer using `inline` instead of `static` here to avoid the duplicate symbol linker error.

rsanthanam-amd commented:
@B1tway 

This looks very good!

I have one minor nitpick so if that is addressed, I can approve this PR.

Thank you!

B1tway commented:
@rsanthanam-amd 
I corrected it to `inline` and added some tests.

I think this PR came out too big, so it would be better to split it into several.
* GCNBuilder and tests
* LoadOp implementation
* StoreOp implementation
* 
If you also think that would be better, then tomorrow I can divide it into several
